nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,637 @@
+
+        elif len(input_masks) == 2 and all(input_masks):
+            input_tensors_shapes = [x.tensor_shape for x in graph.get_input_edges(node)]
+            assert len(input_tensors_shapes[0]) == len(input_tensors_shapes[1])
+            input_shape_len = len(input_tensors_shapes[0])
+            # Join consumed masks
+            left_dim_groups, right_dim_groups = [input_masks[i].dim_groups_map for i in range(2)]  # ignore the 3rd bias
+
+            def _both_dim_blocks_exist(left_idx, right_idx):
+                if left_idx in left_dim_groups or \
+                        right_idx in right_dim_groups:
+                    assert left_idx in left_dim_groups and right_idx in right_dim_groups
+                    return True
+                return False
+
+            output_mask = PropagationMask()
+            # Propagating batch dims
+            for dim in range(input_shape_len - 2):
+                if _both_dim_blocks_exist(dim, dim):
+                    assert len(left_dim_groups[dim]) == 1 and len(
+                        right_dim_groups[dim]) == 1, "multiple groups is not supported"
+                    output_mask.dim_groups_map[dim] = [
+                        PropagationGroup.join_groups(left_dim_groups[dim][0], right_dim_groups[dim][0])
+                    ]
+            # Propagating left rows / right cols
+            for idx, dim in enumerate(range(input_shape_len - 2, input_shape_len)):
+                if dim in input_masks[idx].dim_groups_map:
+                    output_mask.dim_groups_map[dim] = input_masks[idx].dim_groups_map[dim]
+
+            # Close branch
+            if _both_dim_blocks_exist(input_shape_len - 1, input_shape_len - 2):
+                left = left_dim_groups[input_shape_len - 1]
+                right = right_dim_groups[input_shape_len - 2]
+                assert len(left) == 1 and len(right) == 1, "multiple groups is not supported"
+                group = PropagationGroup.join_groups(left[0], right[0])
+                group.close_branch()
+                cls.add_consumer_block(group, node)
+        else:
+            output_mask = node.data.get('output_mask', None)
+
+        node.data['output_mask'] = output_mask
+
+    @classmethod
+    def add_consumer_block(cls, group: PropagationGroup, node: NNCFNode) -> None:
+        """
+        Explicitly adds a consumer block to the given group.
+
+        :param group: the given block group.
+        :param node: node that corresponds to the consumer.
+        """
+        if node.layer_attributes is not None:
+            first_block: PropagationBlock = group.get_blocks()[0]
+            consumer_block = PropagationBlock(
+                MaskProducer(node.node_id),
+                size=first_block.size,
+                offset=first_block.offset,
+                pruning_dimension=1,
+                closed_branches=1
+            )
+            group.add_block(consumer_block)
+
```

Comment:
Should we add consumer blocks for other producers as well? Especially for convolutions?

Reply:
It's added for linear layers. Any others except convolutions?

Reply:
Convolutions will be covered in 107117. JPQD doesn't need them now.